### PR TITLE
Make value concatenation more modular

### PR DESCRIPTION
**Pull request by fehnomenal**
While working on #520 I noticed the concatenation of values from the input might get more and more complex over time.
This is an attempt to make the process at least a little bit more modular.

### CODE REVIEW

**Comment by fehnomenal** [comment]
After thinking about it, I belief having a central registry that maps an ordered tuple of types to a concatenating functions makes more sense.
I will come up with something tomorrow.

**Comment by xiaq** [comment]
Sorry for the late feedback. I applaud your take at extensibility, but using reflect to do double dispatch looks overkill to me.

I think that Python's approach is sufficient: to overload the addition operator in `lhs + rhs`, Python supports a `__radd__` method in addition to `__add__`. If calling `lhs.__add__(rhs)` returns a special `NotImplemented` error Python will try `rhs.__add__(lhs)`. We can do virtually the same thing; define

```go
type Catter interface {
  Cat(interface{}) ({interface{}, error)
}

type RCatter interface {
  RCat(interface{}) ({interface{}, error)
}

var ErrCatNotImplemented = errors.New("cat not implemented")
```

**Comment by xiaq** [comment]
BTW, there is no need for the mutex in the code as long as `Register` is not called in parallel with `Concat`. Concurrent reads of maps are safe.

Also, on second thought, the interface and methods should be called `Concatter`, `Concat`, `RConcat`, `RConcatter`. I was thinking to abbreviate the function name as well to just `Cat`, but that will be very awkward to expose as an Elvish builtin. Simply calling everything `concat` is better.

Also, please move the `Concat` function to `eval/vals`. There is no need to create another package.

**Comment by fehnomenal** [comment]
Thank you for your response. Your proposal would essentially be what I tried in my first commit.
I think it has two drawbacks:
1. There still might be the need for type switches inside the implementations of `Concat`.
2. Builtin types like `string` need a wrapper and special handling.

File: eval/vals/concat.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package vals
+
+import (
+	"errors"
+	"fmt"
+)
+
+type (
```
Comment (xiaq) [inline eval/vals/concat.go]: I would rather not use a `type` block, and have those types documented. But I'm not delaying this PR futher.

**Comment by xiaq** [comment]
Yes you need to do a type switch. This is a style used in most of the functions in `eval/vals`. Go doesn't support advanced type dispatch mechanisms out of box, there is always some overhead involved if you need to do such things.

But I see you implemented this anyway :)

**Pull request merged**